cgi-bin/rp.pm

```diff
@@ -1,414 +1,414 @@
-# Print out to RESULTS the details of the election algorithm, using the
-# information in $log that was returned by rank_candidates.
-sub print_details {
-    (my $log) = @_;
-    print main::RESULTS $log;
-}
-
-1;
+package rp;
+
+# This package computes the CIVS Ranked Pairs and MAM rankings of
+# candidates, by providing an appropriate rank_candidates function.
+# See runoff.pm for a specification of this function.
+
+use CGI qw(:standard);
+use civs_common;
+use strict;
+
+our $mam; # whether MAM tiebreaking is to be used.
+          # Otherwise, uses the deterministic CIVS Ranked Pairs
+          # algorithm.
+
+our @rvh; # Random voter hierarchy; used only for MAM tiebreaking.
+          #   Contains 1's and 0's that encode a total order
+	  #   on the choices.
+our $tiebreak = 0;
+          # whether the RVH was used to order majorities
+our $use_strongest_defeat = 0;
+
+# Note: MAM would pick a random ballot and use it to resolve ties in
+# order_pairs. Here we don't try to achieve a total
+# ordering, but instead group identical-strength pairs into
+# "bunches".
+sub order_pairs {
+    my @pair_a = @{$a};
+    my @pair_b = @{$b};
+    if ($pair_a[2] != $pair_b[2]) {
+	return $pair_b[2] <=> $pair_a[2];
+    } elsif ($pair_a[3] != $pair_b[3]) {
+	return $pair_a[3] <=> $pair_b[3];
+    } elsif ($mam) {
+	my $x = $pair_b[0];
+	my $y = $pair_b[1];
+	my $z = $pair_a[0];
+	my $w = $pair_a[1];
+	$tiebreak = 1;
+	# print pre("Using MAM ordering extension\n");
+	# print pre("$x-$y vs $z-$w\n");
+	# print pre("matrix elements: ". $rvh[$w][$y]. " ". $rvh[$y][$w] . " ".
+		  # $rvh[$x][$z]. " ". $rvh[$z][$x]);
+	if ($w != $y && $rvh[$w][$y] != $rvh[$y][$w]) {
+	    return ($rvh[$w][$y] - $rvh[$y][$w]);
+	}
+	return ($rvh[$x][$z] - $rvh[$z][$x]);
+    } else {
+	return 0;
+    }
+}
+
+sub unparse {
+    if (!defined($_[0])) {
+	return 'undefined';
+    }
+    my @p = @{$_[0]};
+    return "($p[0] vs. $p[1] : $p[2]-$p[3])";
+}
+
+# Compute the transitive closure using the Floyd-Warshall
+# algorithm. This is only used for debugging purposes.
+sub Floyd_Warshall {
+  (my $mref, my $n) = @_;
+  my $update = 0;
+  for (my $k = 0; $k < $n; $k++) {
+    for (my $i = 0; $i < $n; $i++) {
+      for (my $j = 0; $j < $n; $j++) {
+# consider going from i to j via k
+# m_ij = max(m_ij, min(s1,s2))
+	if ($mref->[$i][$k] && $mref->[$k][$j] && !$mref->[$i][$j] &&
+	    $i != $j) {
+	    $mref->[$i][$j] = 1;
+	    # print pre("Update $i, $j");
+	    $update = 1;
+	}
+      }
+    }
+  }
+  return $update;
+}
+
+# TransitiveClosure(m, winner, loser, n) adds the preference
+# for winner over loser to the nxn matrix m and computes the
+# transitive closure (in m). If this creates any new cycles,
+# returns 1, otherwise 0. It uses a worklist algorithm because
+# that is faster than a full Floyd-Warshall when the matrix
+# is already pretty much done.
+sub TransitiveClosure {
+    my ($mref, $winner, $loser, $num_choices) = @_;
```

Comment:
sorry, the diff isn't very useful because my editor decided to also correct the DOS line endings; here's the relevant change

Reply:
This seems like a good idea but would you mind undoing the line terminator changes so the change is localized?

Reply:
sure, no problem